### PR TITLE
Specutils 496

### PR DESCRIPTION
**Pull request by rosteen**
Addresses [Issue #496](https://github.com/astropy/specutils/issues/496) and some points raised in the discussion there. A couple notes:

1) The default exciser function for excise_regions is now true_exciser, which simply deletes the array elements in the input spectral region rather than replacing the values with a linear ramp between the boundary points as before.

2) The linear_exciser function now handles regions with multiple subregions, but the behavior if those subregions overlap with each other might still be a little wonky. For now I added a warning telling the user they should probably use non-overlapping regions rather than a bunch of complex code to handle that edge case.

### CODE REVIEW

**Comment by rosteen** [comment]
Note that test_region_extract.test_linear_excise_invert_from_spectrum was passing in previous builds because both linear_exciser and the test itself were incorrect in ways that canceled each other out to passing. The test has been fixed in this build to do what (I think) it was intended to do in the first place.

File: specutils/manipulation/utils.py [hunk 1]
```diff
@@ -1,26 +1,92 @@
...
+       In the case that ``spectrum`` and ``region`` are not the correct types.
+
+    """
+
+    spectral_axis = spectrum.spectral_axis
+    excise_indices = None
+
+    for subregion in region:
+        # Find the indices of the spectral_axis array corresponding to the subregion
+        wavelengths_in = (spectral_axis >= region.lower) & (spectral_axis < region.upper)
```
Comment (nmearl) [inline specutils/manipulation/utils.py]: I know this isn't particular to this PR, but something we certainly should have caught when this was implemented. The spectral axis is not guaranteed to be in wavelength space, so I think we ought to avoid using the term "wavelengths".

```suggestion
        region_mask = (spectral_axis >= region.lower) & (spectral_axis < region.upper)
```

File: specutils/manipulation/utils.py [hunk 2]
```diff
@@ -1,26 +1,92 @@
...
+        temp_indices = np.nonzero(wavelengths_in)[0]
```
Comment (nmearl) [inline specutils/manipulation/utils.py]:
```suggestion
        temp_indices = np.nonzero(region_mask)[0]
```

File: specutils/manipulation/utils.py [hunk 3]
```diff
@@ -34,56 +100,74 @@ def linear_exciser(spectrum, region):
...
-    wavelengths_in = (wavelengths >= region.lower) & (wavelengths < region.upper)
-    inclusive_indices = np.nonzero(wavelengths_in)[0]
-
-    #
-    # Now set the flux values for these indices to be a
-    # linear range
-    #
-
-    s, e = max(inclusive_indices[0]-1, 0), min(inclusive_indices[-1]+1,
+    wavelengths = spectrum.spectral_axis.copy()
```
Comment (nmearl) [inline specutils/manipulation/utils.py]:
```suggestion
    spectral_axis = spectrum.spectral_axis.copy()
```

File: specutils/manipulation/utils.py [hunk 4]
```diff
@@ -34,56 +100,74 @@ def linear_exciser(spectrum, region):
...
+
+        # Add the uncertainty of the two linear interpolation endpoints in
+        # quadrature and apply to the excised region.
+        if new_uncertainty is not None:
+            new_uncertainty[s:e] = np.sqrt(spectrum.uncertainty[s]**2 + spectrum.uncertainty[e]**2)
 
     # Return a new object with the regions excised.
-    return Spectrum1D(flux=modified_flux*spectrum.flux.unit,
+    return Spectrum1D(flux=modified_flux,
                       spectral_axis=wavelengths,
```
Comment (nmearl) [inline specutils/manipulation/utils.py]:
```suggestion
                      spectral_axis=spectral_axis,
```

File: specutils/manipulation/utils.py [hunk 5]
```diff
@@ -34,56 +100,74 @@ def linear_exciser(spectrum, region):
...
+        # Raise a warning if the SpectralRegion has more than one subregion, since
+        # the handling for this is perhaps unexpected
+        warnings.warn("A SpectralRegion with multiple subregions was provided as "
+             "input. This may lead to undesired behavior with linear_exciser if "
+             "the subregions overlap.",
+             AstropyUserWarning)
+
+    for subregion in region:
+        # Find the indices of the spectral_axis array corresponding to the subregion
+        wavelengths_in = (wavelengths >= subregion.lower) & (wavelengths < subregion.upper)
```
Comment (nmearl) [inline specutils/manipulation/utils.py]:
```suggestion
        region_mask = (spectral_axis >= subregion.lower) & (spectral_axis < subregion.upper)
```

File: specutils/manipulation/utils.py [hunk 6]
```diff
@@ -34,56 +100,74 @@ def linear_exciser(spectrum, region):
...
-def excise_regions(spectrum, regions, exciser=linear_exciser):
+def excise_regions(spectrum, regions, exciser=true_exciser):
     """
-    Apply a convolution based smoothing to the spectrum. The kernel must be one
-    of the 1D kernels defined in `astropy.convolution`.
-
-    This method can be used along but also is used by other specific methods below.
+    Method to replace the flux in the defined regions of the spectrum with
+    interpolated values produced by the function given in the ``exciser``
+    argument.
```
Comment (nmearl) [inline specutils/manipulation/utils.py]: This isn't particularly true in the case of a `true_exciser`. Perhaps something like

```suggestion
    Method to remove or replace the flux in the defined regions of the spectrum 
    depending on the function provided in the ``exciser`` argument.
```

File: specutils/manipulation/utils.py [hunk 7]
```diff
@@ -34,56 +100,74 @@ def linear_exciser(spectrum, region):
...
     spectrum : `~specutils.Spectrum1D`
-        The `~specutils.Spectrum1D` object to which the smoothing will be applied.
+        The `~specutils.Spectrum1D` object to which the excision will be applied.
 
     regions : list of `~specutils.SpectralRegion`
         Each element of the list is a `~specutils.SpectralRegion`. The flux
-        between these wavelengths will be "cut out" using the ``exciser``
-        method.
+        between the lower and upper wavelength of each region will be "cut out"
+        and replaced with interpolated values using the ``exciser`` method.
```
Comment (nmearl) [inline specutils/manipulation/utils.py]:
```suggestion
        between the lower and upper spectral axis value of each region will be "cut out"
        and optionally replaced with interpolated values using the ``exciser`` method.
```

File: specutils/manipulation/utils.py [hunk 8]
```diff
@@ -34,56 +100,74 @@ def linear_exciser(spectrum, region):
...
+        Note that non-overlapping regions should be provided as separate
+        `~specutils.SpectralRegion` objects in this list, not as sub-regions
+        in a single object in the list.
 
     exciser: method
```
Comment (nmearl) [inline specutils/manipulation/utils.py]:
```suggestion
    exciser : function
```

File: specutils/manipulation/utils.py [hunk 9]
```diff
@@ -107,25 +191,28 @@ def excise_regions(spectrum, regions, exciser=linear_exciser):
...
-def excise_region(spectrum, region, exciser=linear_exciser):
+def excise_region(spectrum, region, exciser=true_exciser):
     """
-    Apply a convolution based smoothing to the spectrum. The kernel must be one
-    of the 1D kernels defined in `astropy.convolution`.
-
-    This method can be used along but also is used by other specific methods below.
+    Method to replace the flux in the defined region of the spectrum with
+    interpolated values produced by the function given in the ``exciser``
+    argument.
```
Comment (nmearl) [inline specutils/manipulation/utils.py]:
```suggestion
    Method to remove or replace the flux in the defined regions of the spectrum 
    depending on the function provided in the ``exciser`` argument.
```

File: specutils/tests/test_manipulation.py [hunk 10]
```diff
@@ -9,8 +9,31 @@
...
 from ..spectra import Spectrum1D, SpectralRegion, SpectrumCollection
-from ..manipulation import snr_threshold
-
+from ..manipulation import snr_threshold, excise_regions, linear_exciser
+
+def test_true_exciser():
+    np.random.seed(84)
+    spectral_axis = np.linspace(5000,5100,num=100)*u.AA
+    flux = (np.random.randn(100) + 3) * u.Jy
+    spec = Spectrum1D(flux=flux, spectral_axis = spectral_axis)
```
Comment (nmearl) [inline specutils/tests/test_manipulation.py]:
```suggestion
    spec = Spectrum1D(flux=flux, spectral_axis=spectral_axis)
```
